### PR TITLE
Enable `strict: true` in TypeScript config, refactor as necessary

### PR DESCRIPTION
**Pull request by eemeli**
~This builds on top of #2437 and #2438, so marking as draft & rebasing once they're merged.~ Edit: Done; rebased.

This is a set/sequence of changes that fixes/improves the TS types sufficiently to enable `strict: true`. Mostly the changes are really minor, and should not be observable in the compiled JS code. Some refactoring was needed to avoid patterns that aren't TypeScript-y enough to pass muster.

It may be easiest to review this commit-by-commit (starting from "chore: Specify TS typings better (no JS changes)"), as I've bunched up similar sets of changes. In a couple of cases, fixes were applied to code that may be externally visible during some error conditions, but the effect should be that potential unhandled errors are avoided.

### CODE REVIEW

File: translate/src/modules/machinery/components/ConcordanceSearch.tsx
Comment (Pike) [inline translate/src/modules/machinery/components/ConcordanceSearch.tsx:15]: We should filter on something better here than asserting type, I think.

File: translate/src/modules/resourceprogress/components/ProgressChart.tsx
Comment (Pike) [inline translate/src/modules/resourceprogress/components/ProgressChart.tsx:55]: Just a spoon full of sugar...

```suggestion
    const context = canvas?.getContext('2d');
    if (!context) return;
```

File: translate/src/modules/search/components/SearchBox.test.js
Comment (Pike) [inline translate/src/modules/search/components/SearchBox.test.js:134]: I haven't gotten to review this, but if it's easy for you to disclose, why is it good to remove these tests?

File: translate/src/store.ts
Comment (Pike) [inline translate/src/store.ts:9]: This smells like something we should have a follow-up for?

My gut says we have a typing bug here. In the sense of 1+.

**Review by Pike — COMMENTED** [review]
Not yet at a full review, but figured I should share some notes.

I'm torn on the single-line-if-return. I'm torn for my personal projects. I'm often tempted doing it and then get into an argument between my inner me and my inner stas.

File: translate/src/core/editor/hooks/useUpdateTranslationStatus.ts
Comment (eemeli) [reply translate/src/core/editor/hooks/useUpdateTranslationStatus.ts:48]: I was kind of hoping to be able to avoid the problem later on by getting rid of all redux.

File: translate/src/modules/machinery/components/ConcordanceSearch.tsx
Comment (eemeli) [reply translate/src/modules/machinery/components/ConcordanceSearch.tsx:15]: To be fair, this is a bit on TS for not capturing the result of the filtering always being a `string[]`. As in, it doesn't matter what we `filter()` on, the typing result will still presume it could be null.

File: translate/src/modules/search/components/SearchBox.test.js
Comment (eemeli) [reply translate/src/modules/search/components/SearchBox.test.js:134]: Because they're testing an internal function that shouldn't be public in the first place, and is inlined in the change. These are the lines of code this is testing, in their entirety: https://github.com/mozilla/pontoon/blob/9faea69c202376f3f3515ef665be04a12b61a130/translate/src/modules/search/components/SearchBox.tsx#L189-L193

File: translate/src/store.ts
Comment (eemeli) [reply translate/src/store.ts:9]: How about if I file a catch-all issue after this lands for tracking all remaining tsc & eslint error workarounds? Would be easier afterwards, as then could point at specific lines of code on `master`.

Comment (Pike) [reply translate/src/store.ts:9]: So, I updated VS Code. Which ships TS 4.6, and this error suddenly is pretty omnipresent. That makes it kinda awkward. Only two errors on master (but still two), one is shared, the other is commentTabRef which I think you're fixing on this branch.

I think the problem is that we don't have an app action type. The reducers just take the local action type, and that doesn't align with the app action type.

Found 46 errors in 15 files.

Errors  Files
     3  src/App.tsx:57
     2  src/core/comments/actions.ts:24
     1  src/core/editor/actions.ts:312
     1  src/core/editor/hooks/useUpdateTranslationStatus.ts:44
     1  src/core/l10n/components/AppLocalizationProvider.tsx:39
     1  src/core/plural/actions.ts:23
     1  src/core/plural/components/PluralSelector.tsx:48
     2  src/modules/batchactions/actions.ts:88
     4  src/modules/batchactions/components/BatchActions.tsx:76
     6  src/modules/entitieslist/components/EntitiesList.tsx:144
    14  src/modules/entitydetails/components/EntityDetails.tsx:135
     2  src/modules/history/actions.ts:185
     3  src/modules/navbar/components/Navigation.tsx:46
     3  src/modules/search/components/SearchBox.tsx:184
     2  src/store.ts:9

Comment (eemeli) [reply translate/src/store.ts:9]: Ugh. Yeah, I see that too. On top of #2444 these are down to 37 errors in 12 files with TS 4.6.

As a workaround, in VS Code you can set it to use the local TS version: https://code.visualstudio.com/docs/typescript/typescript-compiling#_using-the-workspace-version-of-typescript

Could we punt all of these to be a separate issue?

**Review by Pike — COMMENTED** [review]
Overall, this looks good.

I've got some nits inline.

I've also struggled to find a pattern in imports and curly braces. @mathjazz , does that matter to you?

And last, I still think we should specify a version of typescript in `package.json` that works. The current version range is covering versions that we know that work and versions that don't.

File: translate/src/modules/search/components/FiltersPanel.tsx
Comment (eemeli) [reply translate/src/modules/search/components/FiltersPanel.tsx:15]: Ah, yeah, this should be just after the `../constants` import.

File: translate/src/core/editor/hooks/useUpdateTranslationStatus.ts
Comment (eemeli) [reply translate/src/core/editor/hooks/useUpdateTranslationStatus.ts:48]: Issue filed: #2454.

File: translate/src/core/resource/actions.ts
Comment (eemeli) [reply translate/src/core/resource/actions.ts:66]: Issue filed: #2454.

File: translate/src/store.ts
Comment (eemeli) [reply translate/src/store.ts:9]: Issue filed: #2453.

File: translate/src/modules/search/components/FiltersPanel.tsx
Comment (eemeli) [reply translate/src/modules/search/components/FiltersPanel.tsx:102]: This exact change actually gets applied in 46e28fa4, as a part of #2444. I'd prefer not to fix it here, as it'd require a manual rebase for the later change.

File: translate/src/modules/search/components/SearchBox.tsx
Comment (eemeli) [reply translate/src/modules/search/components/SearchBox.tsx:278]: This ends up improving quite a bit with the ESLint fixes in #2452, here: https://github.com/mozilla/pontoon/pull/2452/commits/f0b12cd717508a91d7ce3647a366032883892f31#diff-c7a8290c5d0a33d346b9b36743b941153d29629fe24fbf702b729fcf3cd0c56aR226-R248

Is that good enough?

**Comment by eemeli** [comment]
> I've got some nits inline.

These should now all be addressed.

> I've also struggled to find a pattern in imports and curly braces. @mathjazz , does that matter to you?

Added #2452 for this.

> And last, I still think we should specify a version of typescript in `package.json` that works. The current version range is covering versions that we know that work and versions that don't.

While the `package.json` range is rather liberal, `typescript` is a dev dependency and as such we ought to be able to rely on the exact specification in `package-lock.json`.

File: translate/src/modules/search/components/SearchBox.tsx
Comment (Pike) [reply translate/src/modules/search/components/SearchBox.tsx:278]: Yes, thanks.

File: translate/src/modules/search/components/FiltersPanel.tsx
Comment (Pike) [reply translate/src/modules/search/components/FiltersPanel.tsx:100]: That's cool, I had a similar idea of potentially inlining that into the lookup.

**Review by Pike — APPROVED** [review]
Sweet, I think we're good with the follow-ups coming in follow-up PRs.

**Comment by mathjazz** [comment]
Great work! I'll let your merge this, Eemeli.

I've added #2295 to the list of linked issues. Maybe there are other [typescript](https://github.com/mozilla/pontoon/labels/typescript) issues we can close? #2300?

**Comment by eemeli** [comment]
#2300 is not solved here, though it is made more obvious. I rephrased that issue to make it clearer.

**Pull request merged**
